qiita_db/sql_connection.py

```diff
@@ -630,3 +632,381 @@ def get_temp_queue(self):
+                    except IndexError:
+                        # A previous query that was expected to retrieve
+                        # some data from the DB did not return as many
+                        # values as expected
+                        self._raise_execution_error(
+                            sql, sql_args,
+                            "The placeholder {%d:%d:%d} does not match to "
+                            "any previous result"
+                            % (q_idx, r_idx, v_idx))
+                    except TypeError:
+                        # The query that the placeholder is pointing to
+                        # is not expected to retrieve any value
+                        # (e.g. an INSERT w/o RETURNING clause)
+                        self._raise_execution_error(
+                            sql, sql_args,
+                            "The placeholder {%d:%d:%d} is referring to "
+                            "a SQL query that does not retrieve data"
+                            % (q_idx, r_idx, v_idx))
+        return sql, sql_args
+
+    @_checker
+    def add(self, sql, sql_args=None, many=False):
+        """Add an sql query to the transaction
+
+        If the current query needs a result of a previous query in the
+        transaction, a placeholder of the form '{#:#:#}' can be used. The first
+        number is the index of the previous SQL query in the transaction, the
+        second number is the row from that query result and the third number is
+        the index of the value within the query result row.
+        The placeholder will be replaced by the actual value at execution time.
+
+        Parameters
+        ----------
+        sql : str
+            The sql query
+        sql_args : list of objects, optional
+            The arguments to the sql query
+        many : bool, optional
+            Whether or not we should add the query multiple times to the
+            transaction
+
+        Raises
+        ------
+        TypeError
+            If `sql_args` is provided and is not a list
+        RuntimeError
+            If invoked outside a context
+
+        Notes
+        -----
+        If `many` is true, `sql_args` should be a list of lists, in which each
+        list of the list contains the parameters for one SQL query of the many.
+        Each element on the list is all the parameters for a single one of the
+        many queries added. The amount of SQL queries added to the list is
+        len(sql_args).
+        """
+        if not many:
+            sql_args = [sql_args]
+
+        for args in sql_args:
+            if args:
+                if not isinstance(args, list):
+                    raise TypeError("sql_args should be a list. Found %s"
+                                    % type(args))
+            else:
+                args = []
+            self._queries.append((sql, args))
+            self.index += 1
```

Comment:
That makes sense, thanks. It sounds like `index == len(self._queries) +
len(self._results)` at any given time then?

Not tooo concerned about someone setting `index` but agree that would be
one way to mitigate

On Thu, Jul 2, 2015 at 11:07 AM, josenavas <notifications@github.com> wrote:

> In qiita_db/sql_connection.py
> <https://github.com/biocore/qiita/pull/1320#discussion_r33799766>:
>
> > +        Each element on the list is all the parameters for a single one of the
> > +        many queries added. The amount of SQL queries added to the list is
> > +        len(sql_args).
> > +        """
> > +        if not many:
> > +            sql_args = [sql_args]
> > +
> > +        for args in sql_args:
> > +            if args:
> > +                if not isinstance(args, list):
> > +                    raise TypeError("sql_args should be a list. Found %s"
> > +                                    % type(args))
> > +            else:
> > +                args = []
> > +            self._queries.append((sql, args))
> > +            self.index += 1
>
> You're correct, the index is intended for the caller to form the
> placeholders. However, it is not always len(self._queries). The reason is
> transactions can be partially executed but not committed. When that
> happens, we reset self._queries so we don't re-execute them, but
> self._results is kept, and self.index is the one who keeps track of how
> many queries have been inserted (which is not the same as executed).
>
> The only reason why I think it will make sense to have it as property is
> to avoid the consumer of this to modify it. What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/qiita/pull/1320/files#r33799766>.
>